distsys/resources/tcpmailboxes.go

```diff
@@ -0,0 +1,350 @@
+}
+
+var _ distsys.ArchetypeResource = &tcpMailboxesRemoteArchetypeResource{}
+
+func tcpMailboxesRemoteArchetypeResourceMaker(dialAddr string) distsys.ArchetypeResourceMaker {
+	return distsys.ArchetypeResourceMakerFn(func() distsys.ArchetypeResource {
+		return &tcpMailboxesRemoteArchetypeResource{
+			dialAddr: dialAddr,
+		}
+	})
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) ensureConnection() error {
+	if res.conn == nil {
+		var err error
+		res.conn, err = net.Dial("tcp", res.dialAddr)
+		if err != nil {
+			res.conn, res.connEncoder, res.connDecoder = nil, nil, nil
+			fmt.Printf("failed to dial %s, aborting after 50ms: %v", res.dialAddr, err)
+			time.Sleep(tcpMailboxesConnectionDroppedRetryDelay)
+			return distsys.ErrCriticalSectionAborted
+		}
+		res.connEncoder = gob.NewEncoder(res.conn)
+		res.connDecoder = gob.NewDecoder(res.conn)
+	}
+	return nil
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) Abort() chan struct{} {
+	// nothing to do; the remote end tolerates just starting over with no explanation
+	res.inCriticalSection = false // but note to ourselves that we are starting over, so we re-send the begin record
+	return nil
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) PreCommit() chan error {
+	ch := make(chan error, 1)
+	go func() {
+		var err error
+		handleError := func() {
+			fmt.Printf("network error while performing pre-commit handshake, aborting: %v", err)
+			res.conn = nil
+			ch <- distsys.ErrCriticalSectionAborted
+		}
+		// be resilient to somehow reaching this without any sends
+		if !res.inCriticalSection {
+			res.inCriticalSection = true
+			err = res.connEncoder.Encode(tcpNetworkBegin)
+			if err != nil {
+				handleError()
+				return
+			}
+		}
+		err = res.connEncoder.Encode(tcpNetworkPreCommit)
+		if err != nil {
+			handleError()
+			return
+		}
+		var ack struct{}
+		err = res.connDecoder.Decode(&ack)
+		if err != nil {
+			handleError()
+			return
+		}
+		ch <- nil
+	}()
+	return ch
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) Commit() chan struct{} {
+	ch := make(chan struct{}, 1)
+	go func() {
+		var err error
+	outerLoop:
+		for {
+			if err != nil {
+				panic(fmt.Errorf("network error during commit: %s", err))
+			}
+			err = res.ensureConnection()
+			if err != nil {
+				continue outerLoop
+			}
+
+			err = res.connEncoder.Encode(tcpNetworkCommit)
+			if err != nil {
+				continue outerLoop
+			}
+			var shouldResend bool
+			err = res.connDecoder.Decode(&shouldResend)
+			if err != nil {
+				continue outerLoop
+			}
+			if shouldResend {
+				panic("resending is not implemented")
+			}
+			res.inCriticalSection = false
+			ch <- struct{}{}
+			return
+		}
+	}()
+	return ch
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) ReadValue() (distsys.TLAValue, error) {
+	panic(fmt.Errorf("attempted to read from a remote mailbox archetype resource"))
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) WriteValue(value distsys.TLAValue) error {
```

Comment:
Note that this function can be blocking forever. This contradicts the assumption documented in the `ArchetypeResource` interface.